File: tests/network-setups/fabric/dev/scripts/fabric-cli/src/helpers/interop-setup/configure-network.ts
/*
 * Copyright IBM Corp. All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import * as fs from 'fs'
import * as path from 'path'
import {
  invoke,
  getCurrentNetworkCredentialPath,
  getCredentialPath
} from '../fabric-functions'
import { handlePromise, getNetworkConfig } from '../helpers'

const helperInvoke = async (ccFunc, ccArg, ...args) => {
  const [contractName, channelName, connProfilePath, networkName, logger] = args
  const [invokeResponse, invokeError] = await handlePromise(
    invoke(
      {
        contractName,
        channel: channelName,
        ccFunc: ccFunc,
        args: [ccArg]
      },
      connProfilePath,
      networkName,
      global.__DEFAULT_MSPID__,
      logger
    )
  )
  logger.debug(`${ccFunc} Invoke ${JSON.stringify(invokeResponse)}`)
  if (invokeError) {
    logger.error(`${ccFunc} Invoke Error: ${ccArg}`)
    throw new Error(`${ccFunc} Invoke Error ${invokeError}`)
  } else {
    logger.info(`Successfully invoked ${ccFunc}`)
  }
}
const configureNetwork = async (mainNetwork: string, logger: any = console) => {
  const networkEnv = getNetworkConfig(mainNetwork)
  logger.debug(`NetworkEnv: ${JSON.stringify(networkEnv)}`)
  if (!networkEnv.relayEndpoint || !networkEnv.connProfilePath) {
    logger.error(
      'Please use a valid --local-network. If valid network please check if your environment variables are configured properly'
    )
    return
  }

  const credentialFolderPath = getCredentialPath()
  const networkFolders = fs
    .readdirSync(credentialFolderPath, { withFileTypes: true })
    .filter(dirent => dirent.isDirectory())
    .map(item => item.name)
  for (const index in networkFolders) {
    const network = networkFolders[index]
    const accessControlPath = path.join(
      getCurrentNetworkCredentialPath(network),
      'access-control.json'
    )
    const membershipPath = path.join(
      getCurrentNetworkCredentialPath(network),
      'membership.json'
    )
    const verificationPolicyPath = path.join(
      getCurrentNetworkCredentialPath(network),
      'verification-policy.json'
    )
    if (
      !fs.existsSync(accessControlPath) ||
      !fs.existsSync(verificationPolicyPath) ||
      !fs.existsSync(membershipPath)
    ) {
      logger.error(`Missing credential file for network: ${network}`)
    } else {
      await configureNetworkHelper(
        networkEnv.connProfilePath,
        mainNetwork,
        process.env.DEFAULT_CHANNEL ? process.env.DEFAULT_CHANNEL : 'mychannel',
        process.env.DEFAULT_CHAINCODE
          ? process.env.DEFAULT_CHAINCODE
          : 'interop',
        accessControlPath,
        membershipPath,
        verificationPolicyPath,
        logger
      )
    }
  }
}
const configureNetworkHelper = async (
  connProfilePath: string,
  networkName: string,
  channelName: string,
  contractName: string,
  accessControlPath: string,
  membershipPath: string,
  verificationPolicyPath: string,
  logger: any = console
): Promise<void> => {
  const accessControl = Buffer.from(
    fs.readFileSync(accessControlPath)
  ).toString()

  const verificationPolicy = Buffer.from(
    fs.readFileSync(verificationPolicyPath)
  ).toString()

  const membership = Buffer.from(fs.readFileSync(membershipPath)).toString()
  const helperInvokeArgs = [
    contractName,
    channelName,
    connProfilePath,
    networkName,
    logger
  ]
  try {
    await helperInvoke(
      'CreateAccessControlPolicy',
      accessControl,
      ...helperInvokeArgs
    )
  } catch (e) {
    logger.info('CreateAccessControlPolicy attempting Update')
    await helperInvoke(
      'UpdateAccessControlPolicy',
      accessControl,
      ...helperInvokeArgs
    )
  }
  try {
    await helperInvoke(
      'CreateVerificationPolicy',
      verificationPolicy,
      ...helperInvokeArgs
    )
  } catch (e) {
    logger.info('CreateVerificationPolicy attempting Update')
    await helperInvoke(
      'UpdateVerificationPolicy',
      verificationPolicy,
      ...helperInvokeArgs
    )
  }
  try {
    await helperInvoke('CreateMembership', membership, ...helperInvokeArgs)
  } catch (e) {
    logger.info('CreateMembership attempting Update')
    await helperInvoke('UpdateMembership', membership, ...helperInvokeArgs)
  }
}

export { configureNetworkHelper, configureNetwork }

File: weaver/samples/fabric/go-cli/helpers/interopsetup/configure-network.go
/*
Copyright 2020 IBM All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package interopsetup

import (
	"errors"
	"fmt"
	"io/ioutil"
	"path/filepath"

	log "github.com/sirupsen/logrus"
	"github.com/hyperledger-labs/weaver-dlt-interoperability/samples/fabric/go-cli/helpers"
)

// helper functions to log and return errors
func logThenErrorf(format string, args ...interface{}) error {
	errorMsg := fmt.Sprintf(format, args...)
	log.Error(errorMsg)
	return errors.New(errorMsg)
}

func HelperInvoke(ccFunc string, ccArgs []string, contractName, channelName, connProfilePath, networkName string) ([]byte, error) {
	query := helpers.QueryType{
		ContractName: contractName,
		Channel:      channelName,
		CcFunc:       ccFunc,
		Args:         ccArgs,
	}

	result, err := helpers.Invoke(query, connProfilePath, networkName, "Org1MSP", "")
	if err != nil {
		return nil, logThenErrorf("%s helpers.Invoke error: %+v", ccFunc, err.Error())
	}
	log.Infof("%s helpers.Invoke result: %s", ccFunc, result)

	return result, nil
}

func ConfigureNetwork(mainNetwork string) error {
	networkEnv, err := helpers.GetNetworkConfig(mainNetwork)
	if err != nil {
		return logThenErrorf("failure of helpers.GetNetworkConfig for network %s with error: %s", mainNetwork, err.Error())
	}
	log.Infof("network configuration for the network %s is: %+v", mainNetwork, networkEnv)

	if networkEnv.RelayEndPoint == "" || networkEnv.ConnProfilePath == "" {
		return logThenErrorf("please use a valid --local-network. If valid network please check if your environment variables are configured properly")
	}

	credentialFolderPath := helpers.GetCredentialPath()
	fileExists, err := helpers.CheckIfFileOrDirectoryExists(credentialFolderPath)
	if err != nil {
		logThenErrorf("failed to find credentialFolderPath %q: %+s", credentialFolderPath, err.Error())
	} else if !fileExists {
		logThenErrorf("credentials folder %s doesn't exist", credentialFolderPath)
	}

	accessControlPath := filepath.Join(helpers.GetCurrentNetworkCredentialPath(mainNetwork), "access-control.json")
	fileExists, err = helpers.CheckIfFileOrDirectoryExists(accessControlPath)
	if err != nil {
		logThenErrorf("failed to find accessControlPath %q: %+s", accessControlPath, err.Error())
	} else if !fileExists {
		logThenErrorf("file access-control.json doesn't exist")
	}

	membershipPath := filepath.Join(helpers.GetCurrentNetworkCredentialPath(mainNetwork), "membership.json")
	fileExists, err = helpers.CheckIfFileOrDirectoryExists(membershipPath)
	if err != nil {
		logThenErrorf("failed to find membershipPath %q: %+s", membershipPath, err.Error())
	} else if !fileExists {
		logThenErrorf("file membership.json doesn't exist")
	}

	verificationPolicyPath := filepath.Join(helpers.GetCurrentNetworkCredentialPath(mainNetwork), "verification-policy.json")
	_, err = filepath.Glob(verificationPolicyPath)
	if err != nil {
		logThenErrorf("failed to find verificationPolicyPath %q: %+s", verificationPolicyPath, err.Error())
	}
	fileExists, err = helpers.CheckIfFileOrDirectoryExists(verificationPolicyPath)
	if err != nil {
		logThenErrorf("failed to find verificationPolicyPath %q: %+s", verificationPolicyPath, err.Error())
	} else if !fileExists {
		logThenErrorf("file verification-policy.json doesn't exist")
	}

	return configureNetworkHelper(networkEnv.ConnProfilePath, mainNetwork, "mychannel", "interop", accessControlPath, membershipPath, verificationPolicyPath)
}

func configureNetworkHelper(connProfilePath, networkName, channelName, contractName, accessControlPath, membershipPath, verificationPolicyPath string) error {
	accessControlBytes, err := ioutil.ReadFile(filepath.Clean(accessControlPath))
	if err != nil {
		logThenErrorf("failed reading access-control.json with error: %s", err.Error())
	}
	log.Infof("accessControl: %s", string(accessControlBytes))

	ccArgs := []string{string(accessControlBytes)}
	ccFunc := "CreateAccessControlPolicy"
	//ccFunc := "UpdateAccessControlPolicy"
	result, err := HelperInvoke(ccFunc, ccArgs, contractName, channelName, connProfilePath, networkName)
	if err != nil {
		logThenErrorf("%s interopsetup.HelperInvoke error: %+v", ccFunc, err.Error())
	}
	log.Printf("%s interopsetup.HelperInvoke result: %s", ccFunc, string(result))

	membershipBytes, err := ioutil.ReadFile(filepath.Clean(membershipPath))
	if err != nil {
		logThenErrorf("failed reading membership.json with error: %s", err.Error())
	}
	log.Infof("membership: %s", string(membershipBytes))

	ccArgs = []string{string(membershipBytes)}
	ccFunc = "CreateMembership"
	//ccFunc = "UpdateMembership"
	result, err = HelperInvoke(ccFunc, ccArgs, contractName, channelName, connProfilePath, networkName)
	if err != nil {
		logThenErrorf("%s interopsetup.HelperInvoke error: %s", ccFunc, err.Error())
	}
	log.Printf("%s interopsetup.HelperInvoke result: %s", ccFunc, string(result))

	verificationPolicyBytes, err := ioutil.ReadFile(filepath.Clean(verificationPolicyPath))
	if err != nil {
		logThenErrorf("failed reading membership.json with error: %s", err.Error())
	}
	log.Infof("verificationPolicy: %s", string(verificationPolicyBytes))

	ccArgs = []string{string(verificationPolicyBytes)}
	ccFunc = "CreateVerificationPolicy"
	//ccFunc = "UpdateVerificationPolicy"
	result, err = HelperInvoke(ccFunc, ccArgs, contractName, channelName, connProfilePath, networkName)
	if err != nil {
		logThenErrorf("%s interopsetup.HelperInvoke error: %s", ccFunc, err.Error())
	}
	log.Printf("%s interopsetup.HelperInvoke result: %s", ccFunc, string(result))

	return nil
}
